projects/hathor-wallet-headless/hsm-c-bridge.md

```diff
@@ -0,0 +1,345 @@
+
+Since the usage of this kind of device is a niche of large corporate partners, this should be implemented initially in a dedicated `Dockerfile` for HSM-enabled clients, ensuring a reproducible environment for both tests and production code.
+
+### Sample code
+Sample code for signing a transaction:
+```c
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "dinamo.h" /* Dinamo headers */
+#include "htrUtils.h" /* Hathor headers */
+
+int main(int argc, char **argv)
+{
+  // Return value of the latest operation
+  int nRet = 0;
+  struct AUTH_PWD authPwd;
+  HSESSIONCTX hSession = NULL;
+
+  // XPriv Key name
+  char szId[] = HTR_TEST_XPRIV_KEYNAME;
+
+  BYTE ver = DN_BCHAIN_VER_WIF_TEST_NET;
+
+  // Address to derive child priv key from
+  int nChild = 0;
+
+  // Child Priv Key Name
+  char szDest[MAX_OBJ_ID_FQN_LEN] = "HTR_CHILD_KEY";
+  DN_BCHAIN_KEY_INFO pbKeyInfo;
+
+  // Message buffer
+  BYTE pbMsg[1024] = {0};
+
+  // Output encoding buffer
+  char pbOutputEncode[1024 * 2 + 1];
+  DWORD dwOutputLength = sizeof(pbOutputEncode);
+
+  // Hash buffer
+  BYTE pbHash[DN_BCHAIN_HASH_KECCAK256_LEN] = {0};
+  DWORD dwHashLen = sizeof(pbHash);
+
+   // Signature buffer
+  BYTE pbSig[DN_BCHAIN_MAX_SIG_LEN] = {0};
+  DWORD dwSigLen = sizeof(pbSig);
+
+  // Public Key retrieval
+  DN_BCHAIN_PBK pPbk = {0};
+
+  //=======================================================================
+  // Command Line Input Validation example, for alternative implementations
+  //=======================================================================
+  if (argc < 3) {
+    printf("Missing input parameters\n");
+    printf("Usage: signOnPath path {number}\n");
+    return 1;
+  }
+  nChild = atoi(argv[2]);
+
+  // Stdin input implementation, the main way to communicate with the bridge
+  printf("Insert the message to be signed on stdin: ");
+  if (fgets(pbMsg, sizeof(pbMsg), stdin) == NULL) {
+    perror("Error reading message input");
+    return 2;
+  }
+  printf("Received message: %s\n", pbMsg);
+  printf("Will sign this message with path %d\n", nChild);
+
+  //=========================
+  // Initializing Dinamo libs
+  //=========================
+  nRet = DInitialize(0);
+  if (nRet)
+  {
+    printf("Failure at: DInitialize \nError code: %d\n", nRet);
+    exit(1);
+  }
+
+  // Initializing data structure to establishing connection with the HSM
+  strncpy(authPwd.szAddr, HOST_ADDR, sizeof(authPwd.szAddr));
+  authPwd.nPort = DEFAULT_PORT;
+  strncpy(authPwd.szUserId, USER_ID, sizeof(authPwd.szUserId));
+  strncpy(authPwd.szPassword, USER_PWD, sizeof(authPwd.szPassword));
+
+  nRet = DOpenSession(&hSession, SS_USER_PWD, (BYTE *)&authPwd, sizeof(authPwd), ENCRYPTED_CONN);
+  if (nRet)
+  {
+    printf("Failure at: DOpenSession \nError code: %d\n", nRet);
+    goto clean;
```

Comment:
I think we should avoid using goto at all.
Maybe on the actual code the session can be static global and a method would run cleanup.
Then here it would be a simple function call to cleanup state before exiting.